Deploy Magda with Terraform (#5)
magda-io/magda-config: README.md

#### 10. Grant service account permission:

Grant `editor` role to your service account:

```bash
gcloud projects add-iam-policy-binding $PROJECT_ID --member serviceAccount:$SERVICE_ACCOUNT_EMAIL --role roles/editor
```

Grant `k8s admin` role to your service account:
```bash
gcloud projects add-iam-policy-binding $PROJECT_ID --member serviceAccount:$SERVICE_ACCOUNT_EMAIL --role roles/container.admin
```

#### 11. Clone this repo & Initiate Terraform
 and run:

```bash
terraform init
```

After a bit of waiting you should get this message:

```
Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

#### 12. Edit terraform config

> Edit [terraform/magda/terraform.tfvars](./terraform/magda/terraform.tfvars) and supply the follow parameters:

Reviewer: This shouldn't be a link to a file. It should just be in backticks.
```suggestion
Edit `terraform/magda/terraform.tfvars` and supply the follow parameters:
```

Author: Does it give you any troubles to find the file?
It supposes to provide a reference point back to the repository.

Reviewer: The reference to the file in the repository is unnecessary. It's misleading to see a link. But no, it doesn't give me any trouble regarding finding the file.

> Edit [terraform/magda/terraform.tfvars](./terraform/magda/terraform.tfvars) and supply the follow parameters:

Reviewer: Also, the namespace needs to be such that it is not already taken. The default namespace is `default`, so maybe add a note saying something along the lines of "Change the namespace to a new name such that it's not taken."

- `Project id`: the id of the google cloud project that you created (`echo $PROJECT_ID`)
- `Deploy Region`: which region you want to deploy magda to
- `credential_file_path`: the path of the service account key file (`key.json`) that we just generated
- `namespace`: which kubernetes namespace you want to deploy Magda to (generally this should just be "default") 
- `external_domain`: Optional: what domain you want the Magda server to be accessed from (which requires a bit of extra configuration). Leave blank to just access your instance through a temporary domain. You can set this later if necessary.

Other optional settings and their default values (if not set) are:
- `cluster_node_pool_machine_type`: The machine type to use, see https://cloud.google.com/compute/vm-instance-pricing for more details. Default: `n1-standard-4`
- `kubernetes_dashboard`: Whether turn on [kubernetes_dashboard](https://github.com/kubernetes/dashboard) or not; Default: `false`

You can find full list of configurable options from [here](./terraform/magda/variables.tf).

#### 13. Edit default helm config

> Look at [config.yaml](./config.yaml). It has reasonable defaults but you might want to edit something - it will give you a new instance with a standard colour scheme/logos and no datasets (yet).

Reviewer: Again, I think it makes more sense to make it a text and not a link.
```suggestion
Look at `config.yaml`. It has reasonable defaults but you might want to edit something - it will give you a new instance with a standard colour scheme/logos and no datasets (yet).
```

### 2. Install Terraform

Go to [https://learn.hashicorp.com/terraform/getting-started/install.html](https://learn.hashicorp.com/terraform/getting-started/install.html) for instructions

### 3. Install [Google Cloud SDK](https://cloud.google.com/sdk/docs/downloads-interactive)

Go to [https://cloud.google.com/sdk/docs/downloads-interactive](https://cloud.google.com/sdk/docs/downloads-interactive) for instructions.

Once `Google Cloud SDK` is installed, you also need to install gcloud beta components by the following command:

```bash
gcloud components install beta
```

### 4. Create a Google Cloud Project

Before you start the deployment process, you need to create a [google cloud project](https://cloud.google.com/resource-manager/docs/creating-managing-projects) via [Google Cloud Console](https://console.cloud.google.com/) and note down the `Project Id`. Note that this isn't necessarily exactly the same as the id you specified - if it's already been taken, Google will append some numbers to it. Make sure by checking the "Select a Project" dialog in Google Cloud:

![Google Cloud Select a Project Dialog](./gcp-projects.png)

### 5. Set Default Project

> Set the project id you noted down to an environment variable, because you'll need it in a few places - this will work in bash, if you're using another shell either figure out the equivalent or just manually replace $PROJECT_ID with your project id.

Reviewer: ```suggestion
Set the project id you noted down to an environment variable, because you'll need it in a few places - this will work in bash. If you're using another shell use the equivalent command or just manually replace `$PROJECT_ID` with your project id.
```

```bash
export PROJECT_ID=[your-project-id]
```

Then set it as the default in Google Cloud

```bash
gcloud config set project $PROJECT_ID
```

#### 6. Enable required services & APIs for your project

```bash
gcloud services enable compute.googleapis.com
gcloud services enable container.googleapis.com
```

#### 7. Create service account for the deployment

```bash
gcloud iam service-accounts create magda-robot
```

Feel free to use a name other than `magda-robot` if you like.

#### 8. Find out service account email

You need to find out the service account email of your newly created service account to be used as the identifier in other commands.

> To do so, you can firstly, list all service accounts:

Reviewer: ```suggestion
To do so, first list all service accounts:
```

```bash
gcloud iam service-accounts list
```

Find the row of your service account. The service account email should be something similar to `magda-robot@[your-project-id].iam.gserviceaccount.com`. You'll need this a few times, so it's worth saving it to an environment variable - once again, if you're not using a shell that supports this you can just manually replace $SERVICE_ACCOUNT_EMAIL with the email address itself.

```bash
export SERVICE_ACCOUNT_EMAIL=[your-service-account-email]
```

#### 9. Create an access key for your service account

First go to the `terraform/magda` directory inside your cloned version of this repository.

```bash
cd terraform/magda
```

```bash
gcloud iam service-accounts keys create key.json --iam-account=$SERVICE_ACCOUNT_EMAIL
```

You will now have a `key.json` file in `terraform/magda`, containing a private key. We suggest you put this somewhere safe like a password manager.

> DO NOT CHECK IT INTO SOURCE CONTROL.

Reviewer: ```suggestion
DO **NOT** CHECK IT INTO SOURCE CONTROL.
```

#### 10. Grant service account permission

Grant `editor` role to your service account:

```bash
gcloud projects add-iam-policy-binding $PROJECT_ID --member serviceAccount:$SERVICE_ACCOUNT_EMAIL --role roles/editor
```

Grant `k8s admin` role to your service account:

```bash
gcloud projects add-iam-policy-binding $PROJECT_ID --member serviceAccount:$SERVICE_ACCOUNT_EMAIL --role roles/container.admin
```

#### 11. Initiate Terraform

>  and run:

Reviewer: ```suggestion
To do so, run:
```
